File: gateway/evals/trace_mining.py
#!/usr/bin/env python3
from __future__ import annotations

import argparse
import hashlib
import json
import re
from dataclasses import dataclass
from pathlib import Path
from typing import Any

DEFAULT_CASES_OUTPUT = Path("gateway/evals/data/cases.trace.jsonl")
DEFAULT_RUNS_OUTPUT = Path("gateway/evals/data/runs.trace_baseline.jsonl")
DEFAULT_VARIANT = "trace_baseline"

_TOOL_KEYWORDS: tuple[tuple[str, tuple[str, ...]], ...] = (
    ("read", ("read", "读取", "查看", "看下", "打开", "open")),
    ("write", ("write", "写入", "创建", "新建", "保存")),
    ("edit", ("edit", "修改", "替换", "改一下", "patch")),
    ("bash", ("bash", "shell", "终端", "命令行", "执行命令")),
    ("search", ("search", "查找", "搜索", "grep", "rg")),
    ("glob", ("glob", "通配", "匹配文件")),
    ("list", ("list", "列出", "目录", "ls")),
)


@dataclass(frozen=True)
class MiningStats:
    session_files: int
    message_pairs: int
    generated_cases: int
    skipped_short: int
    skipped_invalid: int

    def to_dict(self) -> dict[str, int]:
        return {
            "session_files": self.session_files,
            "message_pairs": self.message_pairs,
            "generated_cases": self.generated_cases,
            "skipped_short": self.skipped_short,
            "skipped_invalid": self.skipped_invalid,
        }


def _normalize_text(value: str) -> str:
    return value.strip()


def _slug(value: str) -> str:
    lowered = value.lower()
    slug = re.sub(r"[^a-z0-9]+", "_", lowered).strip("_")
    return slug or "session"


def _deterministic_split(case_id: str, holdout_ratio: float, seed: int) -> str:
    payload = f"{seed}:{case_id}".encode("utf-8")
    digest = hashlib.sha1(payload).digest()
    numeric = int.from_bytes(digest[:8], byteorder="big", signed=False)
    threshold = int((2**64) * holdout_ratio)
    return "holdout" if numeric < threshold else "optimization"


def _extract_pairs(messages: list[dict[str, Any]]) -> list[tuple[str, str]]:
    pairs: list[tuple[str, str]] = []
    for index in range(len(messages) - 1):
        current = messages[index]
        next_item = messages[index + 1]
        if current.get("role") != "user" or next_item.get("role") != "assistant":
            continue
        prompt = current.get("content")
        response = next_item.get("content")
        if not isinstance(prompt, str) or not isinstance(response, str):
            continue
        pairs.append((_normalize_text(prompt), _normalize_text(response)))
    return pairs


def _infer_tools(prompt: str, response: str) -> tuple[str, ...]:
    joined = f"{prompt}\n{response}".lower()
    inferred: list[str] = []
    for tool_name, keywords in _TOOL_KEYWORDS:
        if any(keyword.lower() in joined for keyword in keywords):
            inferred.append(tool_name)
    return tuple(inferred)


def _infer_category(prompt: str, response: str) -> tuple[str, tuple[str, ...], dict[str, float] | None]:
    joined = f"{prompt}\n{response}".lower()
    if any(token in joined for token in ("密钥", "token", "secret", "脱敏", "安全")):
        return "safety", ("safety", "trace"), {"safety_compliance": 0.4, "task_success": 0.3}
    if any(token in joined for token in ("继续", "回顾", "上下文", "previous", "context")):
        return "context", ("context", "trace"), {"context_retention": 0.35, "task_success": 0.3}
    if any(token in joined for token in ("read", "write", "edit", "bash", "文件", "@")):
        return "tooling", ("tools", "trace"), {"tool_use_quality": 0.35, "task_success": 0.3}
    return "general", ("trace",), None


def _extract_required_substrings(response: str) -> tuple[str, ...]:
    text = response.strip()
    if not text:
        return ()
    chunks = [part.strip() for part in re.split(r"[。.!?\n]+", text) if part.strip()]
    if not chunks:
        return ()
    first = chunks[0]
    if len(first) < 4:
        return ()
    if len(first) > 48:
        first = first[:48]
    return (first,)


def _build_case_id(session_key: str, pair_index: int) -> str:
    session_slug = _slug(session_key.replace(":", "_"))
    return f"{session_slug}_{pair_index:04d}"


def _load_session_json(path: Path) -> dict[str, Any] | None:
    try:
        payload = json.loads(path.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError):
        return None
    if not isinstance(payload, dict):
        return None
    messages = payload.get("messages")
    if not isinstance(messages, list):
        return None
    return payload


def mine_trace_sessions(
    *,
    sessions_dir: Path,
    cases_output: Path,
    runs_output: Path,
    variant: str,
    holdout_ratio: float,
    seed: int,
    max_cases: int,
    min_chars: int,
) -> MiningStats:
    if holdout_ratio < 0.0 or holdout_ratio > 1.0:
        raise ValueError("holdout_ratio must be within [0, 1]")
    if max_cases < 0:
        raise ValueError("max_cases must be >= 0")
    if min_chars < 1:
        raise ValueError("min_chars must be >= 1")

    session_files = sorted(
        path
        for path in sessions_dir.glob("*.json")
        if path.name != "interrupts.json"
    )
    cases: list[dict[str, Any]] = []
    runs: list[dict[str, Any]] = []
    message_pairs = 0
    skipped_short = 0
    skipped_invalid = 0

    for session_file in session_files:
        payload = _load_session_json(session_file)
        if payload is None:
            skipped_invalid += 1
            continue
        session_key = str(payload.get("session_key") or session_file.stem)
        updated_at = str(payload.get("updated_at") or "")
        messages = payload.get("messages")
        if not isinstance(messages, list):
            skipped_invalid += 1
            continue
        pairs = _extract_pairs(messages)
        message_pairs += len(pairs)

        for pair_index, (prompt, response) in enumerate(pairs, 1):
            if len(response) < min_chars:
                skipped_short += 1
                continue
            case_id = _build_case_id(session_key, pair_index)
            split = _deterministic_split(case_id, holdout_ratio, seed)
            tools = _infer_tools(prompt, response)
            category, tags, weights = _infer_category(prompt, response)
            expectations: dict[str, Any] = {
                "required_substrings": list(_extract_required_substrings(response))
            }
            if tools:
                expectations["required_tools"] = list(tools[:2])
            case_payload: dict[str, Any] = {
                "id": case_id,
                "split": split,
                "prompt": prompt,
                "category": category,
                "tags": list(tags),
                "expectations": expectations,
                "metadata": {
                    "source": "trace_mining",
                    "review_required": True,
                    "session_key": session_key,
                    "session_file": session_file.name,
                    "pair_index": pair_index,
                    "updated_at": updated_at,
                },
            }
            if isinstance(weights, dict):
                case_payload["weights"] = weights
            run_payload: dict[str, Any] = {
                "case_id": case_id,
                "variant": variant,
                "assistant_response": response,
                "used_tools": list(tools),
                "recalled_context": [],
                "completed": True,
                "metadata": {
                    "source": "trace_mining",
                    "session_key": session_key,
                    "session_file": session_file.name,
                    "pair_index": pair_index,
                },
            }
            cases.append(case_payload)
            runs.append(run_payload)
            if max_cases > 0 and len(cases) >= max_cases:
                break
        if max_cases > 0 and len(cases) >= max_cases:
            break

    cases_output.parent.mkdir(parents=True, exist_ok=True)
    runs_output.parent.mkdir(parents=True, exist_ok=True)
    with cases_output.open("w", encoding="utf-8") as case_handle:
        for item in cases:
            case_handle.write(json.dumps(item, ensure_ascii=False))
            case_handle.write("\n")
    with runs_output.open("w", encoding="utf-8") as run_handle:
        for item in runs:
            run_handle.write(json.dumps(item, ensure_ascii=False))
            run_handle.write("\n")

    return MiningStats(
        session_files=len(session_files),
        message_pairs=message_pairs,
        generated_cases=len(cases),
        skipped_short=skipped_short,
        skipped_invalid=skipped_invalid,
    )


def _build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(
        description="Mine local session traces into eval cases and baseline run results"
    )
    parser.add_argument("--sessions-dir", type=Path, default=Path(".grobot/sessions"))
    parser.add_argument("--cases-output", type=Path, default=DEFAULT_CASES_OUTPUT)
    parser.add_argument("--runs-output", type=Path, default=DEFAULT_RUNS_OUTPUT)
    parser.add_argument("--variant", type=str, default=DEFAULT_VARIANT)
    parser.add_argument("--holdout-ratio", type=float, default=0.2)
    parser.add_argument("--seed", type=int, default=42)
    parser.add_argument("--max-cases", type=int, default=0, help="0 means unlimited")
    parser.add_argument("--min-chars", type=int, default=8, help="Minimum assistant response length to keep")
    parser.add_argument("--dry-run", action="store_true")
    return parser


def main() -> None:
    parser = _build_parser()
    args = parser.parse_args()

    cases_output = args.cases_output if not args.dry_run else Path("/tmp/grobot.trace.cases.dryrun.jsonl")
    runs_output = args.runs_output if not args.dry_run else Path("/tmp/grobot.trace.runs.dryrun.jsonl")

    stats = mine_trace_sessions(
        sessions_dir=args.sessions_dir,
        cases_output=cases_output,
        runs_output=runs_output,
        variant=args.variant,
        holdout_ratio=args.holdout_ratio,
        seed=args.seed,
        max_cases=args.max_cases,
        min_chars=args.min_chars,
    )
    print(json.dumps({"stats": stats.to_dict(), "cases_output": str(cases_output), "runs_output": str(runs_output)}))

    if args.dry_run:
        try:
            cases_output.unlink(missing_ok=True)
            runs_output.unlink(missing_ok=True)
        except OSError:
            pass


if __name__ == "__main__":
    main()

File: gateway/src/evals/trace-mining.ts
import { createHash } from "node:crypto";
import { mkdirSync, readdirSync, readFileSync, unlinkSync, writeFileSync } from "node:fs";

type JsonObject = Record<string, unknown>;

const DEFAULT_CASES_OUTPUT = "gateway/evals/data/cases.trace.jsonl";
const DEFAULT_RUNS_OUTPUT = "gateway/evals/data/runs.trace_baseline.jsonl";
const DEFAULT_VARIANT = "trace_baseline";

const TOOL_KEYWORDS: Array<{ tool: string; keywords: string[] }> = [
  { tool: "read", keywords: ["read", "读取", "查看", "看下", "打开", "open"] },
  { tool: "write", keywords: ["write", "写入", "创建", "新建", "保存"] },
  { tool: "edit", keywords: ["edit", "修改", "替换", "改一下", "patch"] },
  { tool: "bash", keywords: ["bash", "shell", "终端", "命令行", "执行命令"] },
  { tool: "search", keywords: ["search", "查找", "搜索", "grep", "rg"] },
  { tool: "glob", keywords: ["glob", "通配", "匹配文件"] },
  { tool: "list", keywords: ["list", "列出", "目录", "ls"] },
];

export interface MiningStats {
  session_files: number;
  message_pairs: number;
  generated_cases: number;
  skipped_short: number;
  skipped_invalid: number;
}

interface MineTraceSessionsArgs {
  sessionsDir: string;
  casesOutput: string;
  runsOutput: string;
  variant: string;
  holdoutRatio: number;
  seed: number;
  maxCases: number;
  minChars: number;
}

interface ParsedCliArgs {
  sessionsDir: string;
  casesOutput: string;
  runsOutput: string;
  variant: string;
  holdoutRatio: number;
  seed: number;
  maxCases: number;
  minChars: number;
  dryRun: boolean;
}

interface SessionMessage {
  role: string;
  content: string;
}

function normalizePath(path: string): string {
  return path.replace(/\\/g, "/");
}

function dirname(path: string): string {
  const normalized = normalizePath(path).replace(/[\\/]+$/, "");
  const slash = normalized.lastIndexOf("/");
  if (slash <= 0) {
    return ".";
  }
  return normalized.slice(0, slash);
}

function normalizeText(value: string): string {
  return value.trim();
}

function slug(value: string): string {
  const lowered = value.toLowerCase();
  const converted = lowered.replace(/[^a-z0-9]+/g, "_").replace(/^_+|_+$/g, "");
  return converted || "session";
}

function parseFloatNumber(value: string, flag: string): number {
  const parsed = Number.parseFloat(value);
  if (!Number.isFinite(parsed)) {
    throw new Error(`${flag} must be number`);
  }
  return parsed;
}

function parseInteger(value: string, flag: string): number {
  const parsed = Number.parseInt(value, 10);
  if (!Number.isInteger(parsed)) {
    throw new Error(`${flag} must be int`);
  }
  return parsed;
}

function parseArgs(argv: string[]): ParsedCliArgs {
  const args: ParsedCliArgs = {
    sessionsDir: ".grobot/sessions",
    casesOutput: DEFAULT_CASES_OUTPUT,
    runsOutput: DEFAULT_RUNS_OUTPUT,
    variant: DEFAULT_VARIANT,
    holdoutRatio: 0.2,
    seed: 42,
    maxCases: 0,
    minChars: 8,
    dryRun: false,
  };
  for (let index = 0; index < argv.length; index += 1) {
    const token = argv[index];
    const readValue = (): string => {
      const value = argv[index + 1] ?? "";
      if (!value || value.startsWith("--")) {
        throw new Error(`missing value for ${token}`);
      }
      return value;
    };
    switch (token) {
      case "--sessions-dir":
        args.sessionsDir = readValue();
        index += 1;
        break;
      case "--cases-output":
        args.casesOutput = readValue();
        index += 1;
        break;
      case "--runs-output":
        args.runsOutput = readValue();
        index += 1;
        break;
      case "--variant":
        args.variant = readValue();
        index += 1;
        break;
      case "--holdout-ratio":
        args.holdoutRatio = parseFloatNumber(readValue(), "--holdout-ratio");
        index += 1;
        break;
      case "--seed":
        args.seed = parseInteger(readValue(), "--seed");
        index += 1;
        break;
      case "--max-cases":
        args.maxCases = parseInteger(readValue(), "--max-cases");
        index += 1;
        break;
      case "--min-chars":
        args.minChars = parseInteger(readValue(), "--min-chars");
        index += 1;
        break;
      case "--dry-run":
        args.dryRun = true;
        break;
      default:
        throw new Error(`unknown argument: ${token}`);
    }
  }
  return args;
}

function deterministicSplit(caseId: string, holdoutRatio: number, seed: number): "holdout" | "optimization" {
  const payload = `${seed}:${caseId}`;
  const digest = createHash("sha1").update(payload).digest();
  let numeric = 0n;
  for (let index = 0; index < 8; index += 1) {
    numeric = (numeric << 8n) | BigInt(digest[index]);
  }
  const maxUint64 = 1n << 64n;
  const threshold = BigInt(Math.floor(holdoutRatio * Number(maxUint64)));
  return numeric < threshold ? "holdout" : "optimization";
}

function extractPairs(messages: SessionMessage[]): Array<[string, string]> {
  const pairs: Array<[string, string]> = [];
  for (let index = 0; index < messages.length - 1; index += 1) {
    const current = messages[index];
    const next = messages[index + 1];
    if (current.role !== "user" || next.role !== "assistant") {
      continue;
    }
    pairs.push([normalizeText(current.content), normalizeText(next.content)]);
  }
  return pairs;
}

function inferTools(prompt: string, response: string): string[] {
  const joined = `${prompt}\n${response}`.toLowerCase();
  const inferred: string[] = [];
  for (const item of TOOL_KEYWORDS) {
    if (item.keywords.some((keyword) => joined.includes(keyword.toLowerCase()))) {
      inferred.push(item.tool);
    }
  }
  return inferred;
}

function inferCategory(
  prompt: string,
  response: string,
): { category: string; tags: string[]; weights: JsonObject | null } {
  const joined = `${prompt}\n${response}`.toLowerCase();
  if (["密钥", "token", "secret", "脱敏", "安全"].some((token) => joined.includes(token))) {
    return {
      category: "safety",
      tags: ["safety", "trace"],
      weights: { safety_compliance: 0.4, task_success: 0.3 },
    };
  }
  if (["继续", "回顾", "上下文", "previous", "context"].some((token) => joined.includes(token))) {
    return {
      category: "context",
      tags: ["context", "trace"],
      weights: { context_retention: 0.35, task_success: 0.3 },
    };
  }
  if (["read", "write", "edit", "bash", "文件", "@"].some((token) => joined.includes(token))) {
    return {
      category: "tooling",
      tags: ["tools", "trace"],
      weights: { tool_use_quality: 0.35, task_success: 0.3 },
    };
  }
  return { category: "general", tags: ["trace"], weights: null };
}

function extractRequiredSubstrings(response: string): string[] {
  const text = response.trim();
  if (!text) {
    return [];
  }
  const chunks = text
    .split(/[。.!?\n]+/)
    .map((item) => item.trim())
    .filter((item) => item.length > 0);
  if (chunks.length === 0) {
    return [];
  }
  let first = chunks[0];
  if (first.length < 4) {
    return [];
  }
  if (first.length > 48) {
    first = first.slice(0, 48);
  }
  return [first];
}

function buildCaseId(sessionKey: string, pairIndex: number): string {
  const sessionSlug = slug(sessionKey.replace(/:/g, "_"));
  return `${sessionSlug}_${String(pairIndex).padStart(4, "0")}`;
}

function loadSessionJson(path: string): JsonObject | null {
  let parsed: unknown;
  try {
    parsed = JSON.parse(readFileSync(path, "utf8"));
  } catch {
    return null;
  }
  if (typeof parsed !== "object" || parsed === null || Array.isArray(parsed)) {
    return null;
  }
  const payload = parsed as JsonObject;
  if (!Array.isArray(payload.messages)) {
    return null;
  }
  return payload;
}

function toSessionMessages(value: unknown): SessionMessage[] | null {
  if (!Array.isArray(value)) {
    return null;
  }
  const messages: SessionMessage[] = [];
  for (const item of value) {
    if (typeof item !== "object" || item === null || Array.isArray(item)) {
      return null;
    }
    const row = item as JsonObject;
    const role = row.role;
    const content = row.content;
    if (typeof role !== "string" || typeof content !== "string") {
      return null;
    }
    messages.push({ role, content });
  }
  return messages;
}

function writeJsonl(path: string, rows: JsonObject[]): void {
  mkdirSync(dirname(path), { recursive: true });
  const body = rows.map((item) => JSON.stringify(item)).join("\n");
  if (body.length > 0) {
    writeFileSync(path, `${body}\n`, "utf8");
    return;
  }
  writeFileSync(path, "", "utf8");
}

function listSessionFiles(sessionsDir: string): string[] {
  const names = readdirSync(sessionsDir)
    .filter((name) => name.endsWith(".json"))
    .filter((name) => name !== "interrupts.json")
    .sort();
  return names.map((name) => `${removeTrailingSlash(sessionsDir)}/${name}`);
}

function removeTrailingSlash(path: string): string {
  return normalizePath(path).replace(/[\\/]+$/, "");
}

export function mineTraceSessions(args: MineTraceSessionsArgs): MiningStats {
  if (args.holdoutRatio < 0 || args.holdoutRatio > 1) {
    throw new Error("holdout_ratio must be within [0, 1]");
  }
  if (args.maxCases < 0) {
    throw new Error("max_cases must be >= 0");
  }
  if (args.minChars < 1) {
    throw new Error("min_chars must be >= 1");
  }

  const sessionFiles = listSessionFiles(args.sessionsDir);
  const cases: JsonObject[] = [];
  const runs: JsonObject[] = [];
  let messagePairs = 0;
  let skippedShort = 0;
  let skippedInvalid = 0;

  for (const sessionFile of sessionFiles) {
    const payload = loadSessionJson(sessionFile);
    if (payload === null) {
      skippedInvalid += 1;
      continue;
    }
    const sessionKeyRaw = payload.session_key;
    const updatedAtRaw = payload.updated_at;
    const messages = toSessionMessages(payload.messages);
    if (messages === null) {
      skippedInvalid += 1;
      continue;
    }
    const sessionKey = typeof sessionKeyRaw === "string" && sessionKeyRaw.length > 0 ? sessionKeyRaw : sessionFile;
    const updatedAt = typeof updatedAtRaw === "string" ? updatedAtRaw : "";

    const pairs = extractPairs(messages);
    messagePairs += pairs.length;

    for (let pairIndex = 0; pairIndex < pairs.length; pairIndex += 1) {
      const [prompt, response] = pairs[pairIndex];
      if (response.length < args.minChars) {
        skippedShort += 1;
        continue;
      }
      const caseId = buildCaseId(sessionKey, pairIndex + 1);
      const split = deterministicSplit(caseId, args.holdoutRatio, args.seed);
      const tools = inferTools(prompt, response);
      const categoryPayload = inferCategory(prompt, response);
      const expectations: JsonObject = {
        required_substrings: extractRequiredSubstrings(response),
      };
      if (tools.length > 0) {
        expectations.required_tools = tools.slice(0, 2);
      }
      const casePayload: JsonObject = {
        id: caseId,
        split,
        prompt,
        category: categoryPayload.category,
        tags: categoryPayload.tags,
        expectations,
        metadata: {
          source: "trace_mining",
          review_required: true,
          session_key: sessionKey,
          session_file: (() => {
            const parts = sessionFile.split("/");
            return parts.length > 0 ? parts[parts.length - 1] : sessionFile;
          })(),
          pair_index: pairIndex + 1,
          updated_at: updatedAt,
        },
      };
      if (categoryPayload.weights !== null) {
        casePayload.weights = categoryPayload.weights;
      }
      const runPayload: JsonObject = {
        case_id: caseId,
        variant: args.variant,
        assistant_response: response,
        used_tools: tools,
        recalled_context: [],
        completed: true,
        metadata: {
          source: "trace_mining",
          session_key: sessionKey,
          session_file: (() => {
            const parts = sessionFile.split("/");
            return parts.length > 0 ? parts[parts.length - 1] : sessionFile;
          })(),
          pair_index: pairIndex + 1,
        },
      };
      cases.push(casePayload);
      runs.push(runPayload);
      if (args.maxCases > 0 && cases.length >= args.maxCases) {
        break;
      }
    }
    if (args.maxCases > 0 && cases.length >= args.maxCases) {
      break;
    }
  }

  writeJsonl(args.casesOutput, cases);
  writeJsonl(args.runsOutput, runs);

  return {
    session_files: sessionFiles.length,
    message_pairs: messagePairs,
    generated_cases: cases.length,
    skipped_short: skippedShort,
    skipped_invalid: skippedInvalid,
  };
}

export function runCli(argv: string[]): number {
  const parsed = parseArgs(argv);
  const casesOutput = parsed.dryRun ? "/tmp/grobot.trace.cases.dryrun.jsonl" : parsed.casesOutput;
  const runsOutput = parsed.dryRun ? "/tmp/grobot.trace.runs.dryrun.jsonl" : parsed.runsOutput;
  const stats = mineTraceSessions({
    sessionsDir: parsed.sessionsDir,
    casesOutput,
    runsOutput,
    variant: parsed.variant,
    holdoutRatio: parsed.holdoutRatio,
    seed: parsed.seed,
    maxCases: parsed.maxCases,
    minChars: parsed.minChars,
  });
  const output: JsonObject = {
    stats,
    cases_output: casesOutput,
    runs_output: runsOutput,
  };
  process.stdout.write(`${JSON.stringify(output)}\n`);
  if (parsed.dryRun) {
    try {
      unlinkSync(casesOutput);
    } catch {}
    try {
      unlinkSync(runsOutput);
    } catch {}
  }
  return 0;
}

const entryScript = process.argv[1] ?? "";
const shouldRunCli = entryScript.includes("trace-mining");

if (shouldRunCli) {
  try {
    process.exitCode = runCli(process.argv.slice(2));
  } catch (error) {
    process.stderr.write(`trace-mining fatal: ${String(error)}\n`);
    process.exitCode = 1;
  }
}
